File: Algorithm/FuzzyQLearning_Matlab/env/sense_obs.m
 function senseObs = sense_obs(staticObs, movingObs,  robot)
% sense_obs - Output list of position, calculated distance and direction, radius
% from each obstacle to the robot.
% Output: x y disObs dirObs radObs

[numSObs, ~] = size(staticObs);
[numMObs, ~] = size(movingObs);
senseObs = zeros(numSObs+numMObs, 5);
posRobot = robot(1:2);

for i=1:numSObs
    posObs = staticObs(i,1:2);
    radObs = staticObs(i, 3);
    [dirObs, disObs] = calc_vector(posRobot, posObs);
    senseObs(i,:) = [posObs, disObs, dirObs, radObs];
end

for i=1:numMObs
    posObs = movingObs(i, 1:2);
    radObs = movingObs(i, 4);
    [dirObs, disObs] = calc_vector(posRobot, posObs);
    senseObs(numSObs+i,:) = [posObs, disObs, dirObs, radObs];
end

% Sort the senseObs from nearest to farest (bubble sort)
numObs = numSObs+numMObs;
if numObs > 1
    for i = 0:(numObs-1)
        not_swapped = true;
        for j = 1:(numObs-i-1)
            if senseObs(j,3) > senseObs(j+1,3)
                temp = senseObs(j,:);
                senseObs(j,:) = senseObs(j+1,:);
                senseObs(j+1,:) = temp;
                not_swapped = false;
            end
        end
        if not_swapped
            break
        end
    end
end

File: Algorithm/Qlearning_Python/env/sense_obs.py
import numpy as np
from helper.calc_add_limit_vector import calc_vector, limit_angle


def sense_obs(staticObs, movingObs, robot):
    """ Calculate the distance and direction from each obstacle to the robot """
    numSObs = staticObs.shape[0]
    numMObs = movingObs.shape[0]
    senseObs = np.zeros((numSObs+numMObs, 5), dtype=np.float64)
    posRobot = robot[0:2]

    for i in range(0, numSObs):
        posObs = staticObs[i, 0:2]
        radObs = staticObs[i, 2]
        [dirObs, disObs] = list(calc_vector(posRobot, posObs))
        senseObs[i, :] = [posObs, disObs, dirObs, radObs]

    for i in range(0, numMObs):
        posObs = movingObs[i, 0:2]
        radObs = movingObs[i, 3]
        [dirObs, disObs] = calc_vector(posRobot, posObs)
        senseObs[numSObs+i, :] = [posObs, disObs, dirObs, radObs]
    # return senseObs

    """ Sort the senseObs from nearest to farest (bubble sort) """
    numObs = numSObs+numMObs
    if numObs > 1:
        for i in range(0, numObs):
            not_swapped = True
            for j in range(1, numObs-i):
                if senseObs[j-1, 2] > senseObs[j, 2]:
                    temp = senseObs[j-1, :]
                    senseObs[j-1, :] = senseObs[j, :]
                    senseObs[j, :] = temp
                    not_swapped = False
            if not_swapped:
                break
    return senseObs
